mahiru/components/domain_administrator.py

```diff
@@ -79,142 +110,162 @@ def execute_step(
+            images = self._ensure_images_available(workdir, assets)
 
-                data_containers = self._start_data_containers(
-                        job_id, images, inputs.keys(), output_bases.keys())
+            input_containers = self._start_input_containers(
+                    job_id, images, inputs.keys())
 
-                config = self._create_config_string(
-                        workdir, inputs, step.outputs.keys(), data_containers)
+            output_containers = self._start_output_containers(
+                    job_id, images, output_bases.keys())
 
-                compute_container = self._run_compute_container(
-                        job_id, images['<compute>'], config)
+            config = self._create_config_string(
+                    workdir, inputs, step.outputs.keys(),
+                    dict(**input_containers, **output_containers))
 
-                self._stop_data_containers(
-                        inputs.keys(), step.outputs.keys(), data_containers)
+            compute_container = self._run_compute_container(
+                    job_id, images['<compute>'], config)
 
-                output_images, output_image_files = self._save_output(
-                        job_id, workdir, step.outputs.keys(), data_containers)
+            output_files = dict()  # type: Dict[str, Path]
+            for name in step.outputs:
+                output_files[name] = self._save_output(
+                        workdir, job_id, name, output_containers[name])
 
-                self._store_output_assets(
-                        step, step_subjob, id_hashes,
-                        output_image_files)
-            finally:
-                self._clean_up(
-                        data_containers, compute_container, output_images,
-                        images)
+            return StepResult(output_files, wd)
 
-    def _download_images(
-            self, workdir: Path, inputs: Dict[str, Asset],
-            compute_asset: ComputeAsset, output_bases: Dict[str, Asset]
-            ) -> Dict[str, Path]:
-        """Download required container images.
+        except Exception:
+            rmtree(wd, ignore_errors=True)
+            raise
+
+        finally:
+            if input_containers is not None:
+                self._remove_containers(input_containers.values())
+
+            if compute_container is not None:
+                self._remove_containers([compute_container])
 
-        This downloads the container images for step inputs, compute
-        asset and step outputs from their source asset stores, and
-        saves them as tarballs in the working directory.
+            if assets is not None:
+                for asset in assets.values():
+                    self._free_image(asset.id)
+
+    def _ensure_image_available(self, asset: Asset, workdir: Path) -> Image:
+        """Ensures the asset's image is available in Docker.
```

Comment:
I think next to ensuring that it's there, it also retrieves it? Maybe update the naming/documentation of this function (and it's plural form)?

Reply:
Or is this because in a later PR they are not retrieved anymore?

Reply:
No, they're being retrieved. I'll expand the docstring a bit, I think the function name is long enough as is :-).